internal/dag/policy.go

```diff
@@ -396,55 +396,67 @@ func ingressRetryPolicy(ingress *networking_v1.Ingress, log logrus.FieldLogger)
 	return rp
 }
 
-func ingressTimeoutPolicy(ingress *networking_v1.Ingress, log logrus.FieldLogger) TimeoutPolicy {
+func ingressTimeoutPolicy(ingress *networking_v1.Ingress, log logrus.FieldLogger) RouteTimeoutPolicy {
 	response := annotation.ContourAnnotation(ingress, "response-timeout")
 	if len(response) == 0 {
 		// Note: due to a misunderstanding the name of the annotation is
 		// request timeout, but it is actually applied as a timeout on
 		// the response body.
 		response = annotation.ContourAnnotation(ingress, "request-timeout")
 		if len(response) == 0 {
-			return TimeoutPolicy{
-				ResponseTimeout: timeout.DefaultSetting(),
-				IdleTimeout:     timeout.DefaultSetting(),
+			return RouteTimeoutPolicy{
+				ResponseTimeout:   timeout.DefaultSetting(),
+				IdleStreamTimeout: timeout.DefaultSetting(),
 			}
 		}
 	}
 	// if the request timeout annotation is present on this ingress
 	// construct and use the HTTPProxy timeout policy logic.
-	tp, err := timeoutPolicy(&contour_api_v1.TimeoutPolicy{
+	tp, _, err := timeoutPolicy(&contour_api_v1.TimeoutPolicy{
 		Response: response,
-	})
+	}, 0)
 	if err != nil {
 		log.WithError(err).Error("Error parsing response-timeout annotation, using the default value")
-		return TimeoutPolicy{}
+		return RouteTimeoutPolicy{}
 	}
 
 	return tp
 }
 
-func timeoutPolicy(tp *contour_api_v1.TimeoutPolicy) (TimeoutPolicy, error) {
+func timeoutPolicy(tp *contour_api_v1.TimeoutPolicy, connectTimeout time.Duration) (RouteTimeoutPolicy, ClusterTimeoutPolicy, error) {
 	if tp == nil {
-		return TimeoutPolicy{
-			ResponseTimeout: timeout.DefaultSetting(),
-			IdleTimeout:     timeout.DefaultSetting(),
-		}, nil
+		return RouteTimeoutPolicy{
+				ResponseTimeout:   timeout.DefaultSetting(),
+				IdleStreamTimeout: timeout.DefaultSetting(),
+			}, ClusterTimeoutPolicy{
+				IdleConnectionTimeout: timeout.DefaultSetting(),
+				ConnectTimeout:        0,
```

Comment:
@tsaarni @sunjayBhatia @skriss while upgrade testing from contour-v1.19 to latest v1.22, seeing this issue where ContourConfig file has timeout.connect-timeout set to 5s, but the final envoy config rendered by the xdscache keeps fluctuating certain Cluster's connect_timeout b/w 5s and 2s on any onChanges(). it is expected to be 5s for all dynamic Cluster's connect_timeout.

I think this resetting of Cluster's ConnectTimeout=0 seems to be causing this. Why is this resetting Cluster's ConnectTimeout when a Route's TimeoutPolicy isn't set ? shouldn't this be setting Cluster's ConnectTimeout to passed func arg connectTimeout also ? (just like in the final return below).

Reply:
@ravilr You are correct, thanks for spotting this!